✨ feat: add ComfyUI integration Phase1(RFC-128)

MapleEve opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [x] ✨ feat

#### 🔀 变更说明 | Description of Change
This commit implements RFC-128 Phase 1, adding complete ComfyUI integration to LobeChat with support for multiple AI image generation models.

## Key Features

### Model Support
- **FLUX Series**: Schnell (ultra-fast), Dev (high-quality), Krea-dev (safety-enhanced), Kontext-dev (image editing)
- **Stable Diffusion**: SD 1.5, SDXL, SD 3.5 (Large/Medium), Custom models
- **Image Operations**: Text-to-image and image-to-image generation
- **Flexible Parameters**: Aspect ratios, samplers, schedulers, guidance scale

### Architecture & Implementation
- **Service Layer**: Clean separation of concerns with dedicated service modules
- **Error Handling**: Comprehensive error system with proper error codes and i18n
- **Authentication**: Multi-auth support (none, basic, bearer, custom headers)
- **Model Resolution**: Smart model discovery and variant selection
- **Workflow System**: Modular workflow architecture for different models
- **Caching**: TTL-based cache for model lists and performance optimization

### Testing & Documentation
- **Test Coverage**: 98.41% coverage with comprehensive unit and integration tests
- **Documentation**: Complete development guide in English and Chinese
- **Type Safety**: Full TypeScript support with proper type definitions
- **Mock System**: Unified mock system for consistent testing

### Key Components
- Runtime implementation with SDK integration
- Workflow builders for FLUX and SD models
- Model resolver with automatic variant selection
- Prompt splitter with style keyword support
- Error parser with detailed error mapping
- Authentication system with multiple strategies
- Frontend settings UI with validation
- Backend async task support

This implementation provides a solid foundation for AI image generation in LobeChat, with extensible architecture for future enhancements.
<!-- Thank you for your Pull Request. Please provide a description above. -->

#### 📝 补充信息 | Additional Information
## 代码变更概况

这不是一个小功能，而是完整的ComfyUI运行时集成。代码量说明问题 - 这是一次架构级别的扩展。

### 1.5版本基础（0→1.5阶段成果）
- 基础模型运行时框架
- 简单的provider接口
- 基本错误处理
- 初步的异步支持

### 2.0版本升级（新交付）

## 一、核心运行时扩展

### 服务层架构（packages/model-runtime/src/comfyui/services/）
```
新增5个核心服务：
├── comfyuiClient.ts      # ComfyUI API客户端封装（397行）
├── errorHandler.ts       # 统一错误处理服务（122行）
├── imageService.ts       # 图像生成业务逻辑（214行）
├── modelResolver.ts      # 模型解析与验证（279行）
└── workflowBuilder.ts    # 工作流构建协调器（87行）
```

**从1.5到2.0的提升**：
- **1.5**: 简单的API调用
- **2.0**: 完整的服务分层，职责分离，依赖注入

### 错误处理体系升级（packages/model-runtime/src/comfyui/errors/）
```
7个专用错误类：
├── base.ts               # 基础错误类（ComfyUIInternalError）
├── configError.ts        # 配置错误
├── workflowError.ts      # 工作流错误
├── utilsError.ts         # 工具类错误
├── servicesError.ts      # 服务层错误
├── modelResolverError.ts # 模型解析错误
└── typeGuards.ts         # 类型守卫
```

**从1.5到2.0的提升**：
- **1.5**: 通用错误捕获
- **2.0**: 分层错误体系，精确错误定位，用户友好提示

## 二、工作流引擎（全新能力）

### 工作流实现（packages/model-runtime/src/comfyui/workflows/）
```
5个生产级工作流：
├── flux-dev.ts           # FLUX高质量20步（243行）
├── flux-schnell.ts       # FLUX快速4步（174行）
├── flux-kontext.ts       # FLUX企业28步（299行）
├── sd35.ts              # SD3.5标准流程（275行）
└── simple-sd.ts         # 通用SD适配器（359行）
```

**架构特点**：
- 静态JSON工作流定义，复制粘贴友好
- 动态参数注入策略，而非硬编码
- 智能组件解析，适应不同配置

### 工作流设计巧思
每个工作流采用"配置复制友好"设计：
```typescript
// 静态workflow对象（直接从ComfyUI导出）
const workflow = {
  '1': { class_type: 'DualCLIPLoader', inputs: {...} },
  '2': { class_type: 'UNETLoader', inputs: {...} }
};

// 参数注入（在工作流文件内完成）
workflow['5'].inputs.clip_l = clipLPrompt;
workflow['4'].inputs.width = width;

// PromptBuilder配置
const builder = new PromptBuilder(workflow);
```

## 三、配置与注册表系统

### 模型注册表（config/）
```
├── fluxModelRegistry.ts   # FLUX全系列映射（843行，107个variants）
├── sdModelRegistry.ts     # SD系列映射（508行）
├── modelRegistry.ts       # 统一注册表接口（105行）
└── systemComponents.ts    # 系统组件配置（385行）
```

**生产级特性**：
- 107个FLUX模型variants支持
- 优先级管理（Priority 1-4）
- 智能dtype推荐（default/fp8_e4m3fn/fp16）
- 模型家族识别（FLUX/SD1/SDXL/SD3）

### 智能模型解析
ModelResolverService实现三级缓存策略：
```typescript
class ModelResolverService {
  private modelCache: Map<string, string> = new Map();    // 模型映射缓存
  private vaeCache: string[] | null = null;                // VAE列表缓存
  private componentCache: Map<string, string[]> = new Map(); // 组件缓存
}
```


## 技术实现要点

### 1. 认证系统

- 集成JWT token认证流程
- 支持用户名/密码和token两种认证方式
- 认证状态持久化存储

### 2. 错误处理

- `comfyuiErrorParser.ts` (292行) - 解析ComfyUI特定错误
- 区分网络错误、认证错误、工作流错误
- 用户友好的错误信息展示

### 3. 模型兼容性

- 自动检测服务器可用模型
- 动态工作流生成基于检测结果
- 回退机制处理模型不存在情况

## 测试覆盖

**新增测试文件：**

- 工作流测试：5个测试文件，覆盖各模型工作流
- 工具类测试：7个测试文件，覆盖核心工具函数
- 集成测试：主入口测试1889行，覆盖完整流程

## 文档交付

- **开发文档**: `comfyui-development.mdx`
- **用户文档**: `comfyui.zh-CN.mdx`- 完整使用指南
- **环境变量文档**: 更新模型提供商配置说明

## 配置更新

### Docker支持

4个Dockerfile更新，添加ComfyUI相关环境变量和依赖。

## 架构影响

这次集成扩展了整个模型运行时架构：

- 新增完整的ComfyUI提供商类型定义
- 扩展了异步任务系统支持图像生成  
- 更新了错误处理框架
- 集成了标准参数系统


## 当前技术债：
1. `PromptBuilder` 没有给服务层封装，业务直接调用 SDK
目前的决策原因：因为还不清楚社区想要怎么样快捷的维护自定义的 workflow，后续的 discover 模块怎么发展，是否会接入 workflow template，所以暂时先没做，否则会产生很严重的向后兼容技术债

2. `Client` 没有严格拆分模块

3.  `错误处理` 包装的比较过度，后续可以根据实际流程简化

---

GitHub Issues/Discussions Addressed
This PR completes and extends:

- ✅ Extended https://github.com/lobehub/lobe-chat/discussions/8804 - RFC-128 ComfyUI Integration Phase 1 Complete
- ✅ Extended https://github.com/lobehub/lobe-chat/discussions/411 - Enhanced image generation capabilities
- close https://github.com/lobehub/lobe-chat/issues/673 - ComfyUI integration implementation
- close https://github.com/lobehub/lobe-chat/issues/4265 - FLUX model support
- close https://github.com/lobehub/lobe-chat/issues/2788 - Advanced image generation features
- close https://github.com/lobehub/lobe-chat/issues/8800 - ComfyUI provider configuration
- close https://github.com/lobehub/lobe-chat/issues/7818 - Authentication mechanism improvements
- close https://github.com/lobehub/lobe-chat/issues/8870
<!-- Add any other context about the Pull Request here. -->

tjx666 commented on packages/model-runtime/src/comfyui/services/imageService.ts:
```diff
@@ -0,0 +1,280 @@
+        originalHeight = metadata.height;
+      } else {
+        // Fallback for client-side (should never happen in production)
+        log('Warning: Running in browser environment, cannot read image dimensions');
```
直接 throw error 就完了

tjx666 commented on src/app/[variants]/(main)/image/features/GenerationFeed/BatchItem.tsx:
```diff
@@ -115,6 +116,21 @@ export const GenerationBatchItem = memo<GenerationBatchItemProps>(({ batch }) =>
 
   if (isInvalidApiKey) {
+    // Use custom ComfyUIAuth component for ComfyUI provider
+    if (batch.provider === 'comfyui') {
```
这个针对特定 provider 的展示可能应该加到 src/components/InvalidAPIKey/APIKeyForm/index.tsx

MapleEve replied:
[dbc5ef0](https://github.com/lobehub/lobe-chat/pull/9043/commits/dbc5ef02ce4d1da10484d7bf682347e35995f715)

tjx666 commented:
@MapleEve 

地址和端口号。

<img width="1098" height="654" alt="image" src="https://github.com/user-attachments/assets/9579498c-1859-400c-8dda-550ec634a804" />


<img width="1440" height="900" alt="image" src="https://github.com/user-attachments/assets/ee292a94-19a7-4f12-971c-e78c58e5e4ce" />

MapleEve commented on packages/model-runtime/src/comfyui/index.ts:
```diff
@@ -0,0 +1,72 @@
+ * ComfyUI Runtime implementation
+ * Supports text-to-image and image editing
+ */
+export class LobeComfyUI implements LobeRuntimeAI, AuthenticatedImageRuntime {
```
涉及到从 Comfy 服务器拉图片下来的操作，因为会有认证 header，所以有3个方案：

1. types 新增一个接口专门处理这种情况

2. baseai 接口新增一个认证 可选 record处理

3. 不安全的方案：强制改 _runtime 运行时断言然后在 image 服务注入这个header

目前采用的是最符合写法规范的方案1，方案2会有硬编码类型判断，方案3会用 as any  在处理的时候修改内部 `_runtime`

MapleEve commented on packages/model-bank/src/standard-parameters/index.ts:
```diff
@@ -112,6 +112,24 @@ export const ModelParamsMetaSchema = z.object({
 
+  samplerName: z
+    .object({
+      default: z.string().optional().default('euler'),
```
[1ee9676](https://github.com/lobehub/lobe-chat/pull/9043/commits/1ee967613aad7cb5b3e3d74cc042b88e773e0e57)

MapleEve commented on docs/self-hosting/environment-variables/model-provider.zh-CN.mdx:
```diff
@@ -165,6 +165,97 @@ LobeChat 在部署时提供了丰富的模型服务商相关的环境变量，
 - 默认值：`us-east-1`
 - 示例：`us-east-1`
 
+## ComfyUI
```
[dbc5ef0](https://github.com/lobehub/lobe-chat/pull/9043/commits/dbc5ef02ce4d1da10484d7bf682347e35995f715)

tjx666 commented:
晚上和你微信聊的时候，反倒提醒了我一个一直觉得不太对劲的地方。

- mode-runtime 应该是很轻量，跨平台空运行的
- 你可以阅读目前所有其它 createImage 实现，所有都是采用请求 生图 api 服务的方式生图，因此：
  - 需要把 model-runtime 中 confyui 的 createImage 构建 workflow 的逻辑移到一个接口服务
  - 可以在 src/server/services/comfyui 中实现你的 构建 workflow -> 生成 -> 获取生成图片 等
  - 这个服务可以在 src/server/routers/lambda/comfy 中实现

tjx666 commented:
现在的设计是 createImage 跑在 remote server， 那对于本地用户，没有使用 ngrok 之类的情况下，就没办法拉取到 comfyui 生成的图片。ollama 是不是有同样的问题。 @arvinxx

MapleEve commented:
> 晚上和你微信聊的时候，反倒提醒了我一个一直觉得不太对劲的地方。
> 
> * mode-runtime 应该是很轻量，跨平台空运行的
> * 你可以阅读目前所有其它 createImage 实现，所有都是采用请求 生图 api 服务的方式生图，因此：
>   
>   * 需要把 model-runtime 中 confyui 的 createImage 构建 workflow 的逻辑移到一个接口服务
>   * 可以在 src/server/services/comfyui 中实现你的 构建 workflow -> 生成 -> 获取生成图片 等
>   * 这个服务可以在 src/server/routers/lambda/comfy 中实现

可以，我把服务层和工具层和 config 都挪过去，这样 runtime 就非常扁平了，还不会有 sharp 的问题

arvinxx commented:
> ollama 是不是有同样的问题

有，但这个不用太管，因为 ollama 和其他provider一样只跟 baseUrl 绑定，拉取的 model 是相当于这个 baseUrl 下的 model

MapleEve committed:
✨ feat: add ComfyUI integration with comprehensive model support (RFC-128)

This commit implements RFC-128 Phase 1, adding complete ComfyUI integration to LobeChat with support for multiple AI image generation models.

## Key Features

### Model Support
- **FLUX Series**: Schnell (ultra-fast), Dev (high-quality), Krea-dev (safety-enhanced), Kontext-dev (image editing)
- **Stable Diffusion**: SD 1.5, SDXL, SD 3.5 (Large/Medium), Custom models
- **Image Operations**: Text-to-image and image-to-image generation
- **Flexible Parameters**: Aspect ratios, samplers, schedulers, guidance scale

### Architecture & Implementation
- **Service Layer**: Clean separation of concerns with dedicated service modules
- **Error Handling**: Comprehensive error system with proper error codes and i18n
- **Authentication**: Multi-auth support (none, basic, bearer, custom headers)
- **Model Resolution**: Smart model discovery and variant selection
- **Workflow System**: Modular workflow architecture for different models
- **Caching**: TTL-based cache for model lists and performance optimization

### Testing & Documentation
- **Test Coverage**: 98.41% coverage with comprehensive unit and integration tests
- **Documentation**: Complete development guide in English and Chinese
- **Type Safety**: Full TypeScript support with proper type definitions
- **Mock System**: Unified mock system for consistent testing
- **Edge Runtime Compatibility**: Fixed dynamic imports for production deployment

### Key Components
- Runtime implementation with SDK integration
- Workflow builders for FLUX and SD models
- Model resolver with automatic variant selection
- Prompt splitter with style keyword support
- Error parser with detailed error mapping
- Authentication system with multiple strategies
- Frontend settings UI with validation
- Backend async task support via tRPC

### Secondary Development Optimizations
- Tests now support dynamic model additions without breaking
- Flexible assertions allowing extension of style categories
- Import path corrections for monorepo compatibility

This implementation provides a solid foundation for AI image generation in LobeChat, with extensible architecture for future enhancements and optimized for secondary development scenarios.

# Conflicts:
#	locales/zh-CN/providers.json
#	packages/model-runtime/src/core/ModelRuntime.ts
#	packages/model-runtime/src/index.ts
#	packages/model-runtime/src/runtimeMap.ts
#	packages/model-runtime/src/types/type.ts

MapleEve committed:
:bug: fix: internal api call

MapleEve committed:
✅ test: fix ComfyUI Runtime tests after WebAPI refactoring

- Replace internal service mocks with fetch mocks
- Update tests to verify HTTP requests to WebAPI endpoint
- Mock fetch responses for success and error scenarios
- Fix TypeScript type issues
- Ensure all 22 tests pass with new implementation

MapleEve committed:
🐛 fix: resolve ModelRuntime test import issue by using @lobechat/utils

MapleEve committed:
✨ feat: migrate ComfyUI to providers directory structure

- Move ComfyUI from src/comfyui to src/providers/comfyui
- Add ComfyUI and BFL to ModelProvider enum
- Add BFL to UserKeyVaults interface
- Fix all import paths after rebase with main
- All 22 ComfyUI runtime tests passing

MapleEve committed:
🐛 fix: add authentication to ComfyUI WebAPI route

- Wrap POST handler with checkAuth middleware
- Pass jwtPayload to tRPC caller context
- Fix UNAUTHORIZED error when Runtime calls WebAPI

MapleEve committed:
✨ feat: refactor ComfyUI error handling and cleanup duplicate code

Major changes to ComfyUI error handling architecture and code organization:

Error Handling Refactoring:
- Moved all error detection logic (isNetworkError, isModelError, isWorkflowError, etc.) from frontend to backend errorHandlerService
- Backend now handles all error classification and wraps errors in TRPCError with proper cause field
- Frontend comfyuiErrorParser simplified to only handle HTTP status codes and basic error extraction
- Added SyntaxError detection in backend to identify non-ComfyUI services (HTML responses)
- Error flow: Backend detects → wraps in TRPCError → frontend receives pre-processed error

Code Cleanup:
- Removed duplicate seedGenerator.ts from ComfyUI utils (was copy of @lobechat/utils version)
- Updated all workflow files to import generateUniqueSeeds from @lobechat/utils
- Added number.ts export to packages/utils/src/index.ts to expose generateUniqueSeeds
- Eliminated circular dependency risk between packages

Test Fixes:
- Fixed errorHandler tests to check error.cause instead of error directly (TRPCError wrapping)
- Updated comfyuiErrorParser tests to expect ComfyUIBizError for most cases (simplified frontend)
- Fixed comfyuiClient test to check error.cause for TRPCError wrapped errors
- Fixed provider tests to use toMatchObject instead of toThrow for AgentRuntimeError assertions
- Corrected mock data format in WebAPI error test (message field required)

WebAPI Route:
- Added missing params parameter to checkAuth call in ComfyUI route (TypeScript compliance)
- Fixed authentication flow to properly integrate with middleware

All tests now pass with 100% success rate:
- ComfyUI service: 484/484 tests passing
- Model-runtime ComfyUI: 55/55 tests passing
- ESLint: No errors, only minor style warnings

MapleEve committed:
📝 docs: update ComfyUI development documentation with structural changes

- Update directory structure from src/comfyui/ to src/providers/comfyui/
- Add server-side implementation documentation for src/server/services/comfyui/
- Document new WebAPI routes with authentication middleware
- Add unified error parser documentation
- Update all code examples with correct paths
- Apply changes to both English and Chinese versions

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
♻️ refactor: Optimize and cleanup comfy lambda router

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge remote-tracking branch 'origin/rfc-128' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
:bug:️ fix: building and merge typo

MapleEve committed:
Merge branch 'main' into rfc-128

tjx666 commented on src/app/(backend)/webapi/text-to-image/comfyui/route.ts:
```diff
@@ -0,0 +1,98 @@
+import { NextResponse } from 'next/server';
```
移到 webapi/create-image/comfyui/route ？

text-to-image 显然不合理

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
♻️ refactor: move ComfyUI WebAPI route and fix imports

- Move ComfyUI WebAPI route: text-to-image/comfyui → create-image/comfyui
- Update all references to new API path in docs, tests, and provider code
- Fix ComfyUI settings page import: ../[id] → ../../detail/default
- Optimize ComfyUIParamsSchema to use passthrough() and ComfyUIKeyVault type
- Remove duplicate schema definitions to reduce code redundancy

tjx666 committed:
👷 Adjust mac signing for missing certificates

MapleEve committed:
:bug:️ fix: new backend provider

MapleEve committed:
Merge remote-tracking branch 'upstream/main' into rfc-128

MapleEve committed:
Merge branch 'tj/build/pr-desktop-build-permission' into rfc-128

tjx666 committed:
Merge branch 'main' into rfc-128

MapleEve committed:
🐛 fix: replace btoa with cross-platform base64 utility for Node.js compatibility

- Add cross-platform base64 encoding/decoding utilities in @lobechat/utils
- Replace btoa calls in ComfyUI authentication with createBasicAuthCredentials
- Add comprehensive tests for browser and Node.js environments
- Fix Node.js runtime compatibility for ComfyUI authentication

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve committed:
:bug:️ fix: typo

tjx666 committed:
Merge branch 'main' into rfc-128

tjx666 committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

MapleEve commented on src/server/routers/lambda/comfyui.ts:
```diff
@@ -0,0 +1,111 @@
+  /**
+   * Validate ComfyUI connection
+   */
+  validateConnection: authedProcedure
```
之前是想给 ComfyUIClientService 的内容留前端接口，避免后面重新做

  【tRPC 端点】comfyui.validateConnection (Lambda层)
      ↓ 调用
  【Service方法】clientService.validateConnection() (Service层)
      ↓ 调用
  【底层实现】connectionService.validateConnection() (底层)


实际现在用的是

  imageService.createImage()
      ↓ 直接调用
  【Service方法】clientService.validateConnection() (同一个方法)
      ↓ 调用
  【底层实现】connectionService.validateConnection() (同一个底层)

MapleEve replied:
done in adcbc86f104c9a889f490ef96f6732d54b9fe641

tjx666 commented on docs/development/basic/comfyui-development.mdx:
````diff
@@ -0,0 +1,1009 @@
+
+LobeChat ComfyUI integration uses a four-layer service architecture built around the main `LobeComfyUI` class:
+
+```
````
Code blocks without a declared language should use plaintext; Claude 4 has always had this problem.

tjx666 committed:
Merge branch 'main' into rfc-128

MapleEve committed:
:bug:️ fix: add missing plaintext in doc

MapleEve committed:
:bug:️ remove unused tRPC endpoint according to https://github.com/lobehub/lobe-chat/pull/9043#pullrequestreview-3354507458

MapleEve committed:
✨ feat: add VERCEL_URL support for ComfyUI provider

- Support VERCEL_URL environment variable for Vercel deployments
- Follow same pattern as src/envs/app.ts for URL resolution
- Fallback priority: APP_URL -> VERCEL_URL -> localhost

MapleEve committed:
Merge branch 'main' into rfc-128

tjx666 committed:
Merge branch 'main' into rfc-128

MapleEve committed:
Merge branch 'main' into rfc-128

tjx666 committed:
Merge branch 'main' into rfc-128

tjx666 commented:
Very thanks for your work @MapleEve
